### ISSUE TITLE
Dependency bug with System.Memory in net461

### ISSUE THREAD

**Issue by anurse** [dotnet/corefx#29149]
Repro project: https://github.com/anurse/SpanDependencyBug

Steps:
1. Create `net461` project
2. Reference `System.Memory`
3. Use `Span<T>.IndexOf`

Expected Result: You get the index!

Actual Result:

```
Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Numerics.Vectors, Version=4.1.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies. The system cannot find the file specified.
   at System.SpanHelpers.IndexOf(Char& searchSpace, Char value, Int32 length)
   at System.MemoryExtensions.IndexOf[T](ReadOnlySpan`1 span, T value)
   at SpanDependencyBug.Program.Main(String[] args)
```

I believe this is cause by the fact that System.Memory depends on the following for `net45` (see https://dotnet.myget.org/feed/dotnet-core/package/nuget/System.Memory).

* System.Buffers (>= 4.4.0)
* System.Runtime.CompilerServices.Unsafe (>= 4.5.0-preview3-26416-06)

We are building for `net461` in the sample, but the `net45` dependencies are chosen over the `netstandard2.0` ones because NuGet finds an exact target framework name match (**unless this logic has changed since I last knew how this worked ;)**).

At runtime, `System.Memory` only provides a `netstandard2.0` binary, so it loads that binary, which references `System.Numerics.Vectors`, and thus blows up. System.Memory's references:

```
// netstandard, Version=2.0.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
// System.Numerics.Vectors, Version=4.1.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
// System.Runtime.CompilerServices.Unsafe, Version=4.0.4.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a
// System.Buffers, Version=4.0.2.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51
// mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089
```

**Comment by anurse** [comment]
This is blocking usage of the SignalR Client on .NET Framework targets (specifically tested on .NET 4.6.1 and 4.7.1 in WinForms, but other scenarios will be broken to)

**Comment by anurse** [comment]
cc @weshaggard @ericstj

**Comment by ericstj** [comment]
I believe this is due to how we treat "inbox" on .NET 4.x.  When an assembly version is inbox we assume it is in the unification table.  When an assembly is in the unification table, the runtime chops off the build.revision of the ref and zeros them.  In this case `4.1.2.0` is [listed as inbox](https://github.com/dotnet/corefx/blob/d5b142984587ac3c1b8562a76c83d5dede117289/pkg/Microsoft.Private.PackageBaseline/packageIndex.json#L2970) but it isn't in the unification table so I can see how we might get this wrong.

I haven't had a chance to debug this in detail yet, but I suspect that's what we're hitting.

> blocking usage of the SignalR Client on .NET Framework

This shouldn't be *blocking*: to workaround this please manually reference Vectors for now.

**Comment by anurse** [comment]
> This shouldn't be blocking

Sure, I suppose by blocking I mean that it's very non-obvious to a user that when referencing `Microsoft.AspNetCore.SignalR.Client` they need to also reference `System.Numerics.Vectors` manually.

**Comment by ericstj** [comment]
Got it.  Definitely something that needs to be fixed for 2.1 though.

**Comment by ericstj** [comment]
Just some more background: The buildtools issue has been present ever since 2.0 but is only hit in 2.1 due to versioning.  This is first time we have assembly versions which moved past those that were promoted inbox as part of the .NETStandard2.0 wave.  I'm first examining the right fix in BuildTools and implementing that so that I can diff the results and understand the scope of this issue.  Once I have that we can decide what we want to do for various releases to address it.

**Comment by ericstj** [comment]
Looks like me guess was actually wrong.  It's more nuanced than that.

When we try to minimize package dependencies we look at every possible inbox framework.
We copy the dependencies from the non-inbox to the inbox framework *if some of the dependencies are inbox*.
Then we later remove the copies if we find that they are completely redundant.

In this case what happened was none of the netstandard2.0 dependencies were inbox so no trimmed groups were created, but some of the netstandard1.1 dependencies inbox in net45, so we decided created a net45 dependency group.  This is a problem because as soon as we created the net45 dependency group we changed the resolution for net461.

The fix is don't prematurely optimize when deciding to create new groups since we don't yet have the full view of the set of groups.  So for every TFM copy the dependencies and trim them, then reduce once we have the complete view.
